src/fsharp/service/ItemKey.fs

```diff
@@ -0,0 +1,359 @@
+                writeType vref.Type
+            | _ ->
+                writeString "v$"
+                writeValRef vref
+
+        | Item.UnionCase(info, _) -> 
+            writeString "u$"
+            writeEntityRef info.TyconRef
+            
+        | Item.ActivePatternResult(info, _, _, _) ->
+            writeString "r$"
+            info.ActiveTagsWithRanges
+            |> List.iter (fun (nm, _) ->
+                writeString nm)
+
+        | Item.ActivePatternCase elemRef ->
+            writeString "r$"
+            elemRef.ActivePatternInfo.ActiveTagsWithRanges
+            |> List.iter (fun (nm, _) -> writeString nm)
+
+        | Item.ExnCase tcref ->
+            writeString "e$"
+            writeEntityRef tcref
+
+        | Item.RecdField info ->
+            writeString "d$"
+            writeEntityRef info.TyconRef
+            writeString info.Name
+            writeType info.FieldType
+
+        | Item.AnonRecdField(info, tys, i, _) ->
+            writeString "a$"
+            writeString info.ILTypeRef.BasicQualifiedName
+            tys |> List.iter writeType
+            writeInt32 i
+
+        | Item.NewDef ident ->
+            writeString "n$"
+            writeString ident.idText
+
+        | Item.ILField info ->
+            writeString "l$"
+            writeString info.ILTypeRef.BasicQualifiedName
+            writeString info.FieldName
+
+        | Item.Event info ->
+            writeString "t$"
+            writeString info.EventName
+            writeEntityRef info.DeclaringTyconRef
+
+        | Item.Property(nm, infos) ->
+            writeString "p$"
+            writeString nm
+            infos
+            |> List.iter (fun info -> writeEntityRef info.DeclaringTyconRef)
+
+        | Item.TypeVar(nm, typar) ->
+            writeString "y$"
+            writeString nm
+            writeTypar typar
+
+        | Item.Types(_, [ty]) ->
+            writeType ty
+
+        | Item.UnqualifiedType [tcref] ->
+            writeEntityRef tcref
+
+        | Item.MethodGroup(_, [info], _) 
+        | Item.CtorGroup(_, [info]) ->
+            match info with
+            | FSMeth(_, _, vref, _) ->
+                writeValRef vref
+            | ILMeth(_, info, _) ->
+                info.ILMethodRef.ArgTypes
+                |> List.iter writeILType
+                writeILType info.ILMethodRef.ReturnType
+                writeString info.ILName
+                writeType info.ApparentEnclosingType
+            | _ ->
+                writeString "m$"
+                writeEntityRef info.DeclaringTyconRef
+                writeString info.LogicalName
+
+        | Item.ModuleOrNamespaces [x] ->
+            writeString "o$"
+            x.CompilationPath.DemangledPath
+            |> List.iter (fun x -> 
+                writeString x
+                writeString ".")
+            writeString x.LogicalName
+
+        | Item.DelegateCtor ty ->
+            writeString "g$"
```

Comment:
There are a lot of these strings. Clearly they have some meaning, but it's not clear to me what that meaning is. Would it make sense to factor them into well-named constants?

Reply:
Having these as named constants may be a little more helpful. The reason why they are named this way is to make the key string be unique to the symbol type so key strings don't collide. The "$" is used because it's not possible to have a name with a dollar sign in code for a name.

Reply:
I'll make these named constants so there isn't ambiguity and may look clearer when reading the code.